designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,170 @@
+
+As more projects adopt this standardized module format, the hope is a lot of this "connective glue" will fall away, leaving us with leaner modules and less build tooling across our stack.
+
+
+## Detailed Design
+
+### Nomenclature
+
+* `ESM` ecmascript module. The standardized module format for javascript
+* `CJS` commonjs. Node's historical module format (i.e., `module.exports = ...`, `const a = require('foo')`)
+* `UMD` universal module definition. A legacy format for specifying javascript code as a module in a way that tries to guarantee compatibility with many loader formats including requireJS, commonjs, and global script tags.
+* `IIFE` immediately invoked function expressions. A legacy strategy for encapsulating javascript into a module that won't leak global variables into the execution environment. It is still the primary way of packaging global script tags for inclusion in a web page. `UMD` is usually wrapped in an `IIFE`
+* `CDN` content delivery network. can be thought of as a set of servers that host static content all around the globe to deliver assets quickly and at scale.
+
+
+### build process
+
+Today, espree is written in commonjs (CJS) format, and a universal module definition (UMD) build is created too.
+
+
+today's build process:
+```
+┌-------------------┐               ┌-----------------┐
+│                   │               │                 │                package.json:
+│     espree.js     |               │ build/espree.js │
+│ (CJS entry point) ├──BUILD_STEP──▶│   (UMD bundle)  │     CJS ---▶   "main": "espree.js",
+│                   │               │                 │
+└-------------------┘               └-----------------┘
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it.
+
+
+proposed build process in this RFC:
+```
+                                  ┌-------------------┐
+                                  │                   │
+                                  │  dist/espree.cjs  │
+                                  │ (CJS entry point) │                package.json:
+┌-------------------┐             │                   │
+│                   │             └--▲----------------┘     CJS ---▶   "main": "dist/espree.cjs",
+│     espree.js     │                │                                 "exports": {
+│ (ESM entry point) ├───BUILD_STEP───┘                      ESM ---▶     "import": "espree.js",
+│                   │                                       CJS ---▶     "require": "./dist/espree.cjs"
+└-------------------┘                                                  },
+                                                                       "type": "module"
```

Comment:
type module is not required whatsoever; `.mjs` files are always ESM and `.cjs` files are always CJS, and type module makes `.js` mean ESM instead of CJS. for back compat (as well as ideological correctness), however, you actively *do not want* to change how `.js` files are parsed, so a dual package should either omit `type` or set it to `"commonjs"`.

Reply:
> for back compat (as well as ideological correctness), however, you actively _do not want_ to change how `.js` files are parsed, so a dual package should either omit `type` or set it to `"commonjs"

`dist/espree.cjs` will be the only CJS module we publish in v8. All `.js` files we're publishing as CJS in v7 will become ESM in v8, so we want them to be treated as ESM.

Not sure what's the plan for the files we use for development, like [tests](https://github.com/eslint/espree/tree/master/tests) and [tools](https://github.com/eslint/espree/tree/master/tools).

`

Reply:
Right, but why not publish ESM as `.mjs`, rather than needing to change what `.js` has always meant?

Reply:
> rather than needing to change what .js has always meant?

I mean we're getting into opinion territory for sure, but I suspect that `.mjs` will eventually looked back upon as "oh yeah, that was that weird file format proposal that node and google were advocating, back when the module ecosystem was super fragmented".

I think it's inevitable that what `.js` means is changing, that in the somewhat near future it will be assumed to be esm, and not the other way around.

Reply:
That may be true in the future, but that is not true now, and it seems like eslint should match the present instead of a possible future.

Reply:
> That may be true in the future, but that is not true now

I think that's a fair point 👍 

> eslint should match the present instead of a possible future.

I think it's a bit of a balancing act. It obviously needs to match present needs and expectations, but it also needs to prepare for a future that is coming pretty soon. esm has already shipped in browsers, in node, in deno.

Reply:
Sure, but browsers have no concept of extensions at all, and node is highly unlikely to ever change the default parsing behavior, so there's really no downside in sticking to the status quo for what extensions mean.